### PR TITLE
feat: Enable address converting for deployment args

### PR DESCRIPTION
**Pull request by RetricSu**

### CODE REVIEW

File: packages/base/src/process.ts [hunk 1]
```diff
@@ -85,6 +88,94 @@ export async function buildEstimateGasTransaction(
...
+  tx: EthTransaction,
+  signingMethod: any,
+  signingMessageType_?: SigningMessageType
+): Promise<SerializeSignedTransactionString> {
+  const process: Process = {
+    type: ProcessTransactionType.send,
+    signingMethod: signingMethod,
+    signingMessageType: signingMessageType_,
+  };
+  if (formalizeEthToAddress(tx.to) != DEFAULT_EMPTY_ETH_ADDRESS) {
```
Comment (classicalliu) [inline packages/base/src/process.ts:112]:
```suggestion
  if (formalizeEthToAddress(tx.to) !== DEFAULT_EMPTY_ETH_ADDRESS) {
```

File: packages/base/src/process.ts [hunk 2]
```diff
@@ -85,6 +88,94 @@ export async function buildEstimateGasTransaction(
...
+    receiverScriptHash,
+    signingMessageType === SigningMessageType.withPrefix
+  );
+
+  const _signature = await process.signingMethod!(message);
+  const signature = godwoker.packSignature(_signature);
+  const l2Tx = { raw: rawL2Tx, signature: signature };
+
+  const result = splitByteCodeAndConstructorArgs(tx.data, deploymentRecords);
+  if (result == undefined) {
```
Comment (classicalliu) [inline packages/base/src/process.ts:155]:
```suggestion
  if (result == null) {
```

File: packages/base/src/util.ts [hunk 3]
```diff
@@ -1268,7 +1280,274 @@ export class Godwoker {
...
+// then the contract deployment is considered interesting.
+export interface DeploymentRecords {
+  [signature: string]: AbiItem;
+}
+
+export function calculateByteCodeSignature(bytecode: string) {
+  const byteCodeSignature = keccak256(Buffer.from(bytecode, "hex"))
+    .slice(0, CONTRACT_BYTE_CODE_HASH_HEAD_IN_BYTE)
+    .toString("hex");
+  return byteCodeSignature;
```
Comment (classicalliu) [inline packages/base/src/util.ts:1296]: `0x` prefix missing ?

File: packages/base/src/util.ts [hunk 4]
```diff
@@ -1268,7 +1280,274 @@ export class Godwoker {
...
+    return false;
+  }
+
+  const bytecode = inputData.slice(0, byteCodeLength);
+  const bytecodeSignature = calculateByteCodeSignature(bytecode);
+  if (bytecodeSignature != result.byteCodeSignature) {
+    return false;
+  }
+
+  return true;
```
Comment (classicalliu) [inline packages/base/src/util.ts:1368]:
```suggestion
  return bytecodeSignature === result.byteCodeSignature
```

File: packages/base/src/util.ts [hunk 5]
```diff
@@ -1268,7 +1280,274 @@ export class Godwoker {
...
+  if (!abiItem.inputs) {
+    return {
+      newArgs: args,
+      isInterested: false,
+      addressMapping,
+      deploymentRecords,
+    };
+  }
+
+  if (args.length != abiItem.inputs.length) {
```
Comment (classicalliu) [inline packages/base/src/util.ts:1429]:
```suggestion
  if (args.length !== abiItem.inputs.length) {
```

File: packages/base/src/util.ts [hunk 6]
```diff
@@ -1268,7 +1280,274 @@ export class Godwoker {
...
+  return {
+    newArgs: convertArgs,
+    isInterested: true,
+    addressMapping,
+    deploymentRecords,
+  };
+}
+
+export function validateEthAddress(value: any) {
+  if (typeof value != "string") {
```
Comment (classicalliu) [inline packages/base/src/util.ts:1502]:
```suggestion
  if (typeof value !== "string") {
```

File: packages/base/src/util.ts [hunk 7]
```diff
@@ -1268,7 +1280,274 @@ export class Godwoker {
...
+    throw new Error(`ethAddress is not string, ${value}`);
+  }
+
+  if (value.length != 42) {
```
Comment (classicalliu) [inline packages/base/src/util.ts:1506]:
```suggestion
  if (value.length !== 42) {
```

File: packages/base/src/util.ts [hunk 8]
```diff
@@ -1268,7 +1280,274 @@ export class Godwoker {
...
+    throw new Error(`ethAddressArray is not a array, ${value}`);
+  }
+
+  for (let item of value) {
+    validateEthAddress(item);
+  }
+}
+
+export function validateBytecode(value: any) {
+  if (typeof value != "string") {
```
Comment (classicalliu) [inline packages/base/src/util.ts:1526]:
```suggestion
  if (typeof value !== "string") {
```

File: packages/base/src/util.ts (hunk 3 above)
Comment (RetricSu) [reply packages/base/src/util.ts:1296]: > `0x` prefix missing ?

no need for "0x", https://github.com/nervosnetwork/polyjuice-provider/blob/enable-deploy-args/packages/base/src/util.ts#L1307

Comment (classicalliu) [reply packages/base/src/util.ts:1296]: As an exported function, doesn't it has the same principles as other hex string?

Comment (RetricSu) [reply packages/base/src/util.ts]: > As an exported function, doesn't it has the same principles as other hex string?

you are right, we should make it more consistant. updated.

**Review by classicalliu — APPROVED** [review]

**Pull request merged**